src/EditorFeatures/Test/CodeRefactorings/CodeRefactoringServiceTest.cs

```diff
@@ -229,6 +230,63 @@ public async Task TestAnalyzerConfigDocumentRefactoringAsync()
         Assert.Equal(refactoring2.Title, globalConfigRefactoringTitle);
     }
 
+    [Fact]
+    public void TestDelayedLoading()
+    {
+        var composition = FeaturesTestCompositions.Features.AddParts(
+            typeof(NonSourceFileRefactoringWithDocumentKindsAndExtensions),
+            typeof(NonSourceFileRefactoringWithDocumentKinds),
+            typeof(NonSourceFileRefactoringWithDocumentExtensions),
+            typeof(NonSourceFileRefactoringWithoutDocumentKindsAndExtensions));
+
+        using var workspace = TestWorkspace.CreateCSharp("", composition: composition);
+        var refactoringService = (CodeRefactorings.CodeRefactoringService)workspace.GetService<ICodeRefactoringService>();
+
+        var project = workspace.CurrentSolution.Projects.Single()
+            .AddAdditionalDocument("test.TXT", "", filePath: "test.TXT").Project
+            .AddAdditionalDocument("test", "", filePath: "test").Project
+            .AddAdditionalDocument("test.log", "", filePath: "test.log").Project
+            .AddDocument("test.cs", "", filePath: "test.cs").Project
+            .AddDocument("test.editorconfig", "", filePath: "test.editorconfig").Project;
+
+        VerifyProviders(refactoringService,
+            project.AdditionalDocuments.Single(t => t.Name == "test.TXT"),
+            typeof(NonSourceFileRefactoringWithDocumentKindsAndExtensions),
+            typeof(NonSourceFileRefactoringWithDocumentKinds));
+
+        VerifyProviders(refactoringService,
+            project.AdditionalDocuments.Single(t => t.Name == "test"),
+            typeof(NonSourceFileRefactoringWithDocumentKinds));
+
+        VerifyProviders(refactoringService,
+            project.AdditionalDocuments.Single(t => t.Name == "test.log"),
+            typeof(NonSourceFileRefactoringWithDocumentKinds));
+
+        VerifyProviders(refactoringService,
+            project.Documents.Single(t => t.Name == "test.editorconfig"),
+            typeof(NonSourceFileRefactoringWithDocumentExtensions),
+            typeof(NonSourceFileRefactoringWithoutDocumentKindsAndExtensions));
+
+        VerifyProviders(refactoringService,
+            project.Documents.Single(t => t.Name == "test.cs"),
+            typeof(NonSourceFileRefactoringWithoutDocumentKindsAndExtensions));
+    }
+
+    private static void VerifyProviders(CodeRefactorings.CodeRefactoringService service, TextDocument document, params Type[] expectedProviderTypes)
+    {
+        // Exclude provides which have not been setup by test
+        var assembly = Assembly.GetExecutingAssembly();
+        var actualProviders = service.GetProviders(document)
+            .Where(p => p.GetType().Assembly == assembly)
+            .ToArray();
+
+        foreach (var type in expectedProviderTypes)
+        {
+            Assert.Contains(actualProviders, p => p.GetType() == type);
```

Comment:
We have an AssertEx.SetEqual() helper for doing set comparisons.

Reply:
Actual set are CodeRefactorerProvider instances. Expected set are the types of those instances.

Reply:
You could say AssertEx.SetEqual(expectedProviderTypes, actualProviders.Select(p => p.GetType()) then.

(leave this as is though unless you need to do a follow-up).